graphql.md

```diff
@@ -0,0 +1,177 @@
+# GraphQL
+
+Mutations:
+
+- [Name should begin with action type](#name-should-begin-with-action-type)
+- [Use an input object type for the argument](#use-an-input-object-type-for-the-argument)
+- [Return an object in the payload](#return-an-object-in-the-payload)
+- [Errors](#errors)
+
+## Mutations
+
+Schema in the following examples
+
+```
+type User {
+  firstName: String!
+  lastName: String!
+  age: Int!
+}
+```
+
+### Name should begin with action type
+
+✅
+```gql
+
+mutation CreateUserMutation($input: UserInputType!) {
+   createUser(input: $input) {
+       user {
+           firstName
+       }
+   }
+}
+```
+
+✅
+```gql
+
+mutation DeleteUserMutation($input: UserInputType!) {
+   deleteUser(input: $input) {
+       user {
+           firstName
+       }
+   }
+}
+```
+
+### Use an input object type for the argument
+
+✅
+```gql
+
+mutation UpdateUserMutation($input: UserInputType!) {
+   updateUser(input: $input) {
+       user {
+           firstName
+       }
+   }
+}
+```
+
+where 
+
+```
+type UserInputType {
+    firstName: String!
+    lastName: String!
+    age: Int!
+}
+
+```
+
+❌
+```gql
+mutation UpdateUserMutation($firstName: String!, $lastName: String!, $age: Int!) {
+   updateUser(firstName: $firstName, lastName: $lastName, age: $age) {
+       user {
+           firstName
+       }
+   }
+}
+```
+
+Otherwise the mutations become more cumbersome to write out as the number of arguments grow
+
+Input object types are more flexible to changes. If the arguments change, less places need to be updated
+
+In the case where different mutations have the same arguments, the input object type can be reused.
+
+```gql
+mutation CreateUserMutation($input: UserInputType!) {
+   createUser(input: $input) {
+       user {
+           firstName
+       }
+   }
+}
+```
+
+### Return an object in the payload
+
+Ideally the object you've mutated
+
+✅
+```gql
+mutation UpdateUserMutation($input: UserInputType!) {
+   updateUser(input: $input) {
+       user {
+           firstName
+           lastName
+       }
+   }
+}
+```
+
+❌
+```gql
+mutation UpdateUserMutation($input: UserInputType!) {
+   updateUser(input: $input) {
+        firstName
+        lastName
+   }
+}
+```
+
+### Errors
+
+All errors should be returned as top-level errors as stated in the [graphql specification](https://spec.graphql.org/June2018/#sec-Response-Format)
+
+#### Validation errors
+
+These are defined by the `errorClass` `VALIDATION`
+Should return a `validationErrors` field in `extensions`.
+Each error should contain the path to the input which caused the error.
```

Comment:
👏

Will this convention "just work" with validation errors affecting items in a list? E.g. if I have:

```
meterReadings: [
  { serialNumber: "", readAt: "", value: ""},
  ...
]
```

and there's an error at `meterReadings[2].readAt`, then would `inputPath: ["meterReadings", 2, "readAt"]` be what we expect to see in `validationErrors`?

Reply:
yep I'd expect a number to indicate which item it's referring to. 
Just for completeness, I'd expect the mutation variables to look like 
```json
{
    "input": {
        "meterReadings": [{ serialNumber: "", readAt: "BAD INPUT", value: ""}]
    }
}
```
and the error input path
```
inputPath: ["input", "meterReadings", 0, "readAt"]
```

Reply:
Haven't thought to much about list inputs so not sure what the drawbacks could be yet